autoarray/structures/grids.py

```diff
@@ -1555,7 +1715,172 @@ def extent(self):
+            mask_2d=grid.mask,
+            unmasked_sparse_grid_pixel_centres=unmasked_sparse_grid_pixel_centres,
+        )
+
+        sparse_for_unmasked_sparse = sparse_util.sparse_for_unmasked_sparse_from_mask_2d_and_pixel_centres(
+            mask_2d=grid.mask,
+            unmasked_sparse_grid_pixel_centres=unmasked_sparse_grid_pixel_centres,
+            total_sparse_pixels=total_sparse_pixels,
+        ).astype(
+            "int"
+        )
+
+        unmasked_sparse_for_sparse = sparse_util.unmasked_sparse_for_sparse_from_mask_2d_and_pixel_centres(
+            total_sparse_pixels=total_sparse_pixels,
+            mask_2d=grid.mask,
+            unmasked_sparse_grid_pixel_centres=unmasked_sparse_grid_pixel_centres,
+        ).astype(
+            "int"
+        )
+
+        regular_to_unmasked_sparse = grid_util.grid_pixel_indexes_1d_from_grid_scaled_1d_shape_2d_and_pixel_scales(
+            grid_scaled_1d=grid,
+            shape_2d=unmasked_sparse_shape,
+            pixel_scales=pixel_scales,
+            origin=origin,
+        ).astype(
+            "int"
+        )
+
+        sparse_1d_index_for_mask_1d_index = sparse_util.sparse_1d_index_for_mask_1d_index_from_sparse_mappings(
+            regular_to_unmasked_sparse=regular_to_unmasked_sparse,
+            sparse_for_unmasked_sparse=sparse_for_unmasked_sparse,
+        ).astype(
+            "int"
+        )
+
+        sparse_grid = sparse_util.sparse_grid_from_unmasked_sparse_grid(
+            unmasked_sparse_grid=unmasked_sparse_grid_1d,
+            unmasked_sparse_for_sparse=unmasked_sparse_for_sparse,
+        )
+
+        return GridSparse(
+            sparse_grid=sparse_grid,
+            sparse_1d_index_for_mask_1d_index=sparse_1d_index_for_mask_1d_index,
+        )
+
+    @classmethod
+    def from_total_pixels_grid_and_weight_map(
+        cls, total_pixels, grid, weight_map, n_iter=1, max_iter=5, seed=None
+    ):
+        """Calculate the image-plane pixelization from a grid of coordinates (and its mask).
+
+        See *grid_stacks.GridSparse* for details on how this grid is calculated.
+
+        Parameters
+        -----------
+        grid : grids.Grid
+            The grid of (y,x) scaled coordinates at the centre of every image value (e.g. image-pixels).
+        """
+
+        if total_pixels > grid.shape[0]:
+            raise exc.GridException
+
+        kmeans = KMeans(
+            n_clusters=total_pixels, random_state=seed, n_init=n_iter, max_iter=max_iter
+        )
+
+        try:
+            kmeans = kmeans.fit(X=grid.in_1d_binned, sample_weight=weight_map)
+        except ValueError or OverflowError:
+            raise exc.InversionException()
+
+        return GridSparse(
+            sparse_grid=kmeans.cluster_centers_,
+            sparse_1d_index_for_mask_1d_index=kmeans.labels_.astype("int"),
+        )
+
+    @property
+    def total_sparse_pixels(self):
+        return len(self.sparse)
+
+
+class GridTransformed(Grid):
+
```

Comment:
A very useful class

Reply:
There was a nasty bug where grids were transformed multiple time, as the grid_like decorator was losing their TrasnformedGrid class.

This is a hacky fix but a good solution needs a refactor of mask.py, which is gonna be a huge undertaking.